subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/MavenComponentParser.java

```diff
@@ -233,130 +235,94 @@ private static boolean dependencyMatchesProject(ModuleDependency dependency, Mod
+            this.dependencyResolver = dependencyResolver;
+            this.scope = scope;
+            this.optional = optional;
+            this.globalExcludes = globalExcludes;
         }
 
-        private List<MavenDependency> asDependencies(ModuleDependency dependency, String scope, boolean optional, Set<ExcludeRule> globalExcludes) {
+        private void convertDependency(ModuleDependency dependency, Consumer<MavenDependency> collector) {
             Set<ExcludeRule> allExcludeRules = getExcludeRules(globalExcludes, dependency);
+            VariantDependencyResolver.Coordinates coordinates = dependencyResolver.resolveVariantCoordinates(dependency, warnings);
 
             if (dependency.getArtifacts().isEmpty()) {
-                Coordinates coordinates = resolveCoordinates(dependency.getGroup(), dependency.getName(), dependency.getVersion(), null);
-                return Collections.singletonList(newDependency(coordinates, null, null, scope, allExcludeRules, optional));
+                collector.accept(newDependency(coordinates, null, null, scope, allExcludeRules, optional));
+                return;
             }
 
-            List<MavenDependency> dependencies = new ArrayList<>();
             for (DependencyArtifact artifact : dependency.getArtifacts()) {
-                Coordinates coordinates = resolveCoordinates(
-                    dependency.getGroup(),
-                    artifact.getName(), // TODO: This seems wrong. We should not allow the artifact to change the dependency coordinates.
-                    dependency.getVersion(),
-                    null
-                );
-
-                dependencies.add(newDependency(coordinates, artifact.getType(), artifact.getClassifier(), scope, allExcludeRules, optional));
-            }
-            return dependencies;
-        }
-
-        private MavenDependency asDependency(ProjectDependency dependency, String scope, boolean optional, Set<ExcludeRule> globalExcludes) {
-            Path identityPath = getIdentityPath(dependency);
-            Coordinates coordinates = resolveCoordinates(identityPath);
-            Set<ExcludeRule> allExcludeRules = getExcludeRules(globalExcludes, dependency);
-
-            return newDependency(coordinates, null, null, scope, allExcludeRules, optional);
-        }
-
-        private MavenDependency asDependencyConstraint(DependencyConstraint dependency) {
-            Coordinates coordinates = resolveCoordinates(dependency.getGroup(), dependency.getName(), dependency.getVersion(), null);
+                VariantDependencyResolver.Coordinates artifactCoordinates = coordinates;
+                if (!artifact.getName().equals(coordinates.getName())) {
+                    // TODO: We should not allow the artifact name to change the coordinates.
+                    //  Artifacts with name different from the coordinate name is not supported in Maven.
+                    //  This behavior should be deprecated.
```

Comment:
💭 Why not add a deprecation warning when this situation is detected now as part of this change?

Reply:
Wanted to avoid behavior changes and more unrelated code in this PR. Its already at almost lines. I can remove this, though I felt that its at least good to commit the intention of deprecating this so we don't forget to later

Reply:
Making the change in a followup PR is probably better.